File: v25_iterator/python/book2.py
from collections.abc import Iterator

class Book:
    def __init__(self, title):
        self.title = title

    def get_title(self):
        return self.title

class Bookshelf:
    def __init__(self):
        self.books = []

    def add_book(self, book):
        self.books.append(book)

    def iterator(self):
        return BookIterator(self.books)

class BookIterator(Iterator):
    def __init__(self, books):
        self.index = 0
        self.books = books

    def __next__(self):
        try:
            book = self.books[self.index]
        except IndexError:
            raise StopIteration()
        self.index += 1
        return book

def main():
    bookshelf = Bookshelf()
    bookshelf.add_book(Book("Book 1"))
    bookshelf.add_book(Book("Book 2"))
    bookshelf.add_book(Book("Book 3"))

    iterator = bookshelf.iterator()
    for book in iterator:
        print(book.get_title())

if __name__ == "__main__":
    main()

File: v25_iterator/java/Book2.java
package v25_iterator.java;

import java.util.ArrayList;

interface Iterator {
    boolean hasNext();
    Object next();
}

class Book {
    private String title;

    public Book(String title) {
        this.title = title;
    }

    public String getTitle() {
        return title;
    }
}

class Bookshelf {
    private ArrayList<Book> books;

    public Bookshelf() {
        books = new ArrayList<>();
    }

    public void addBook(Book book) {
        books.add(book);
    }

    public Iterator iterator() {
        return new BookIterator();
    }

    private class BookIterator implements Iterator {
        int index;

        public boolean hasNext() {
            return index < books.size();
        }

        public Object next() {
            if (this.hasNext()) {
                return books.get(index++);
            }
            return null;
        }
    }
}

public class Book2 {
    public static void main(String[] args) {
        Bookshelf bookshelf = new Bookshelf();
        bookshelf.addBook(new Book("Book 1"));
        bookshelf.addBook(new Book("Book 2"));
        bookshelf.addBook(new Book("Book 3"));

        Iterator iterator = bookshelf.iterator();
        while (iterator.hasNext()) {
            Book book = (Book) iterator.next();
            System.out.println(book.getTitle());
        }
    }
}
